swap/swap.go

```diff
@@ -1,98 +1,636 @@
+	}
+
+	return actualAmount, nil
+}
+
+// verifyChequeProperties verifies the signature and if the cheque fields are appropriate for this peer
+// it does not verify anything that requires knowing the previous cheque
+func (s *Swap) verifyChequeProperties(cheque *Cheque, p *Peer) error {
+	if cheque.Contract != p.contractAddress {
+		return fmt.Errorf("wrong cheque parameters: expected contract: %x, was: %x", p.contractAddress, cheque.Contract)
+	}
+
+	// the beneficiary is the owner of the counterparty swap contract
+	if err := cheque.VerifySig(p.beneficiary); err != nil {
+		return err
+	}
+
+	if cheque.Beneficiary != s.owner.address {
+		return fmt.Errorf("wrong cheque parameters: expected beneficiary: %x, was: %x", s.owner.address, cheque.Beneficiary)
+	}
+
+	if cheque.Timeout != 0 {
+		return fmt.Errorf("wrong cheque parameters: expected timeout to be 0, was: %d", cheque.Timeout)
+	}
+
+	return nil
+}
+
+// verifyChequeAgainstLast verifies that serial and amount are higher than in the previous cheque
+// furthermore it cheques that the increase in amount is as expected
+// returns the actual amount received in this cheque
+func verifyChequeAgainstLast(cheque *Cheque, lastCheque *Cheque, expectedAmount uint64) (uint64, error) {
+	actualAmount := cheque.Amount
+
+	if lastCheque != nil {
+		if cheque.Serial <= lastCheque.Serial {
+			return 0, fmt.Errorf("wrong cheque parameters: expected serial larger than %d, was: %d", lastCheque.Serial, cheque.Serial)
+		}
+
+		if cheque.Amount <= lastCheque.Amount {
+			return 0, fmt.Errorf("wrong cheque parameters: expected amount larger than %d, was: %d", lastCheque.Amount, cheque.Amount)
+		}
+
+		actualAmount -= lastCheque.Amount
+	}
+
+	if expectedAmount != actualAmount {
+		return 0, fmt.Errorf("unexpected amount for honey, expected %d was %d", expectedAmount, actualAmount)
+	}
+
+	return actualAmount, nil
+}
+
+// To be called with mutex already held
+// Caller must be careful that the same balance isn't concurrently read and written by multiple routines
+func (s *Swap) updateBalance(peer enode.ID, amount int64) (int64, error) {
 	//adjust the balance
 	//if amount is negative, it will decrease, otherwise increase
-	s.balances[peer.ID()] += amount
+	balance, exists := s.getBalance(peer)
+	if !exists {
+		return 0, fmt.Errorf("peer %v does not exist", peer)
+	}
+	newBalance := balance + amount
+	s.setBalance(peer, newBalance)
 	//save the new balance to the state store
-	peerBalance := s.balances[peer.ID()]
-	err = s.stateStore.Put(peer.ID().String(), &peerBalance)
+	err := s.store.Put(balanceKey(peer), &newBalance)
+	if err != nil {
+		return 0, err
+	}
+	log.Debug("balance for peer after accounting", "peer", peer.String(), "balance", strconv.FormatInt(newBalance, 10))
+	return newBalance, err
+}
 
-	log.Debug(fmt.Sprintf("balance for peer %s: %s", peer.ID().String(), strconv.FormatInt(peerBalance, 10)))
-	return err
+// loadBalance loads balances from the state store (persisted)
```

Comment:
load and get are the wrong way round. 
these accessors should be defined on the swapPeer.
loading balance, cheques from store should be done when the peer is initialised


```
p := s.getPeer(id)
p.lastSentCheque()
p.setLastSentCheque()
p.lastReceivedCheque()
p.setLastReceivedCheque()
p.balance()
p.setBalance()
```

Reply:
I think that is the plan, I think we have an issue for this, right @ralph-pichler ?

I vote we postpone this for after the merge.

Reply:
relevant issue: #1604.